sdk/go/common/resource/plugin/plugin_test.go

```diff
@@ -89,3 +97,92 @@ func TestParsePort(t *testing.T) {
 		require.Error(t, err)
 	}
 }
+
+func TestHealthCheck(t *testing.T) {
+	t.Parallel()
+
+	startServer := func(healthService bool) (*grpc.Server, *plugin) {
+		listener, _ := net.Listen("tcp", "127.0.0.1:0")
+		server := grpc.NewServer()
+
+		if healthService {
+			healthServer := health.NewServer()
+			grpc_health_v1.RegisterHealthServer(server, healthServer)
+		}
+
+		go func() {
+			err := server.Serve(listener)
```

Comment:
Shouldn't we waitgroup that this serve has ran before returning from startServer otherwise the healthcheck might fail just because we haven't started up yet?

Reply:
dialPlugin may be doing that on L125. If it responds it must've stared

```
			// The connection is supposedly ready; but we will make sure it is *actually* ready by sending a dummy
			// method invocation to the server.  Until it responds successfully, we can't safely proceed.
			for {
				handshakeRes, err = handshake(timeout, bin, prefix, conn)

```

https://github.com/pulumi/pulumi/blob/bd516f6535105497ee5596c6c2eeea871b1539f8/sdk/go/common/resource/plugin/plugin.go#L189

Reply:
Went ahead and used a channel to make sure the server is ready.